File: kast/kast.py
# https://docs.python.org/release/2.7.2/library/ast.html#abstract-grammar
# https://docs.python.org/release/3.4.2/library/ast.html#abstract-grammar

from ast import *
import ast
import sys
import _ast

# class Args(list):
#     pass

def args():
   return ast.arguments(args=[Name(id='self',ctx=Load())],defaults=[],vararg=None,kwarg=None)

class Todo(ast.expr):
    pass
# (self.)call[dot[1..3]]
# Subscript(Name('x', Load()), Slice(Num(1), Num(3), None), Load())

class Return(ast.Return):
    def __init__(self, **kwargs):
        self.value=None

    def set_val(self,val):
        if isinstance(val,list):val=val[0] #REALLY!!?
        self.value=val
    body = property(lambda self:self.value, set_val)
    variable = property(lambda self:self.value, set_val)

class TryExcept(ast.TryExcept):
    def __init__(self, **kwargs):
        self.orelse=None
        self.handlers=[]

    def set_rescue(self,r):
        self.handlers.append(r)
    rescue = property(lambda self:self.handlers, set_rescue)

class ExceptHandler(ast.ExceptHandler):
    def __init__(self, **kwargs):
        self.type=None
#         name


class arguments(ast.arguments):
    def __init__(self, **kwargs):
        self.args=[] #[Name(id='self',ctx=Load())]
        self.defaults=[]
        self.kwarg=self.vararg=None
        super(ast.arguments,self).__init__(*kwargs)

    def set_args(self,oas):
        if not isinstance(oas,list):oas=[oas]
        if not oas[0]: return
        self.args=self.args+oas
    body = property(lambda self:self.args, set_args)

class Num(ast.Num):
    def set_value(self,n):self.n=n
    value = property(lambda self:self.n, set_value)

    def __repr__(self): return str(self.n)

class Name(ast.Name):
    def set_name(self,id):
        if(isinstance(id,ast.Name)):
            id=id.id #WWWTTTFFF
        self.id=id
    name=property(lambda self:self.id,set_name)

    def __eq__(self, other):
        return self.id==other or self.id==other.id

class Str(ast.Str):
    def set_value(self,s):
        if isinstance(s,ast.Name):
            s=s.id
        self.s=s
    value = property(lambda self:self.s, set_value)
    name=property(lambda self:self.s,set_value)
    def __str__(self):return s

class For(ast.For):
    def __init__(self, **kwargs):
        self.orelse=None

    def set_assign(self,target):
        self.target=target
    assign = property(lambda s:s.target,set_assign)
    def set_iter(self,iter):
        self.iter=iter
    call= property(lambda s:s.iter,set_iter)

class Assign(ast.Assign):
    def set_var(self,var):
        if not isinstance(var,list):var=[var]
        self.targets=var
    def set_val(self,val):
        if isinstance(val,list):val=val[0] #REALLY!!?
        self.value=val
    var = property(lambda self:self.targets, set_var)
    object = property(lambda self:self.targets, set_var)
    name = property(lambda self:self.targets, set_var)
    # variable = property(lambda self:self.targets, set_var)
    # variables = property(lambda self:self.targets, set_var)
    # body = property(lambda self:self.targets, set_var)
    body = property(lambda self:self.value, set_val)
    variable = property(lambda self:self.value, set_val)#REALLY???
    variables = property(lambda self:self.value, set_val)

# a.split('b') =>
# Call(func=Attribute(value=Name(id='a', ctx=Load()), attr='split', ctx=Load()), args=[Str(s='b')]
class Call(ast.Call):
    def __init__(self, **kwargs):
        self.args=[]
        self.keywords=[]
        self.kwargs=self.starargs=None
        self.func=None #Attribute(value=Name(id='self',ctx=Load()),attr=None,ctx=Load())
        super(ast.Call,self).__init__(*kwargs)

    def set_method(self,method):
        if isinstance(method,list):
            method=method[0] #HOW?
        if(isinstance(self.func,Attribute)):
            self.func.attr=method
        else:
            self.func=method
            # self.func=Attribute(value=Name(id='self',ctx=Load()),attr=method,ctx=Load())

    method = property(lambda self:self.func, set_method)
    const = property(lambda self:self.func, set_method)
    name = property(lambda self:self.func, set_method)

    # self.x() Call(func=Attribute(value=Name('self', Load()), attr='x', Load()))
    def set_object(self,o):
        if(not isinstance(self.func,Attribute)):
            self.func=Attribute(attr=self.func,ctx=Load(),value=o)
        self.func.value=o
    def get_object(self):
        if(isinstance(self.func,Attribute)):return self.func.value
        return None
    object = property(get_object, set_object)
    variable = property(get_object, set_object)

    def set_args(self,oas):
        if not isinstance(oas,list):oas=[oas]
        # if isinstance(oas[0],arguments):oas=[oas[0].value] #lolwat??
        if not oas[0]: return
        self.args=self.args+oas
    body = property(lambda self:self.args, set_args)
    # method = property(lambda self:self.func, lambda self,method:[0 for self.func in [method]])

class ClassDef(ast.ClassDef):
    def __init__(self, **kwargs):
        self.nl=True
        self.decorator_list=[]
        self.bases=[]
        super(ast.ClassDef,self).__init__(*kwargs)

# todo : more beautiful mappings / defaults
class Import(ast.Import):
    def set_package(self,package):
        self.names=[package]
    def get_package(self):
        return self.names
    package = property(get_package , set_package )

class FunctionDef(ast.FunctionDef):
    def __init__(self, **kwargs):
        self.decorator_list=[]
        self.body=[Pass()]
        self.args=args()
        super(ast.FunctionDef,self).__init__(*kwargs)

class Print(ast.Print):
      def __init__(self, **kwargs):
          super(ast.Print,self).__init__(*kwargs)
          if not "nl" in kwargs:
              self.nl=True


if sys.version_info > (3,0):
     # PYTHON 3 HACK!!
    class Print(stmt):
        _fields = (
            'dest',
            'values',
            'nl',
        )
    # | Exec(expr body, expr? globals, expr? locals)
    class Exec(Call):
        _fields = (
            'body',
            'globals',
            'locals',
        )


# class Name(ast.Name):
#     # def __init__(self, **kwargs):
#     #     super(ast.Name,self).__init__(*kwargs)
#     def __str__(self):
#         if not 'id' in self._attributes: self.id="MISSING_ID!!!"
#         return "<kast.Name id='%s'>"%self.id

types={ # see _ast.py , F12:
"Raise":Raise,#danger raise keyword
"class":ClassDef,
"operator":operator,
"Add":Add,
"alias":alias,
"boolop":boolop,
"And":And,
"arguments":arguments,
"stmt":stmt,
"Assert":Assert,
"Assign":Assign,
"expr":expr,
"Attribute":Attribute,
"AugAssign":AugAssign,
"expr_context":expr_context,
"AugLoad":AugLoad,
"AugStore":AugStore,
"BinOp":BinOp,
"BitAnd":BitAnd,
"BitOr":BitOr,
"BitXor":BitXor,
"BoolOp":BoolOp,
"Break":Break,
"Call":Call,
"ClassDef":ClassDef,
"cmpop":cmpop,
"Compare":Compare,
"comprehension":comprehension,
"Continue":Continue,
"Del":Del,
"Delete":Delete,
"Dict":Dict,
"DictComp":DictComp,
"Div":Div,
"slice":slice,
"Ellipsis":Ellipsis,
"Eq":Eq,
"excepthandler":excepthandler,
"ExceptHandler":ExceptHandler,
"Exec":Exec,
"Expr":Expr,
"mod":mod,
"Expression":Expression,
"ExtSlice":ExtSlice,
"FloorDiv":FloorDiv,
"For":For,
"FunctionDef":FunctionDef,
"GeneratorExp":GeneratorExp,
"Global":Global,
"Gt":Gt,
"GtE":GtE,
"If":If,
"IfExp":IfExp,
"Import":Import,
"ImportFrom":ImportFrom,
"In":In,
"Index":Index,
"Interactive":Interactive,
"unaryop":unaryop,
"Invert":Invert,
"Is":Is,
"IsNot":IsNot,
"keyword":keyword,
"Lambda":Lambda,
"List":List,
"ListComp":ListComp,
"Load":Load,
"LShift":LShift,
"Lt":Lt,
"LtE":LtE,
"Mod":Mod,
"Module":Module,
"Mult":Mult,
"Name":Name,
"Not":Not,
"NotEq":NotEq,
"NotIn":NotIn,
"Num":Num,
"Or":Or,
"Param":Param,
"Pass":Pass,
"Pow":Pow,
"Print":Print,
'Raise':Raise,
"Repr":Repr,
"Return":Return,
"RShift":RShift,
"Set":Set,
"SetComp":SetComp,
"Slice":Slice,
"Store":Store,
"Str":Str,
"Sub":Sub,
"Subscript":Subscript,
"Suite":Suite,
"TryExcept":TryExcept,
"TryFinally":TryFinally,
"Tuple":Tuple,
"UAdd":UAdd,
"UnaryOp":UnaryOp,
"USub":USub,
"While":While,
"With":With
    }

# workaround: alias is keyword in ruby!
mapped_types={
    "Arguments":arguments,
    # "Block":Module, #NOT REALLY!
    "Variable": Name,
    "Const": Name, #todo
    "Class":ClassDef,
    # "Alias":alias, Assign a=b
    "Argument":Name,
    # "Default":
    "Hash":Dict,
    "String":expr, # str or binop('a',add,'b')
    "Begin":TryExcept,
    "Defs":Todo, # FunctionDef(name='?', args=args(),body=[],decorator_list=[Name(id='classmethod', ctx=Load())]),

    # "Arguments":arguments,
    # "Args":arguments,
    # "Args":List, # Args, AAARRG!!! ;)

    # "class_method":FunctionDef, # FunctionDef(name='x', args=arguments(args=[], vararg=None, kwarg=None, defaults=[]) WTF
    "Method":FunctionDef,
    "int":Num,
    "let":Assign,
    "Symbol":Str, # Name, a: b => 'a':b !!
    "Literal":Name,
    "Condition":expr,
    "Dot":Todo,
    "Regexp":Str, #! r'.' == "." !!!
    "Rescue":ExceptHandler,
    "Self":Name(id='self',ctx= Load()),
    "Super":Call,
    "Yield":Todo, # Yield, #NOOO ;)
    "Value":expr,
    "Then":expr #Value
}


types.update(mapped_types)
for k,v in types.items():
    if(k=='Raise'):continue
    if(k=='Let'):continue
    types[k.lower()]=types[k]
    types["{http://angle-lang.org}"+k.lower()]=types[k]

assert Name(id='xyz',ctx=Load())=='xyz'

File: kast/kast.rb
# https://docs.python.org/release/2.7.2/library/ast.html#abstract-grammar
# https://docs.python.org/release/3.4.2/library/ast.html#abstract-grammar
# -- ASDL's six builtin types are identifier, int, string, bytes, object, >>> singleton <<< NEW

# require_relative '../extensions.rb'

module Kast

  class Node
    attr_accessor :lineno, :col_offset, :children

    def initialize(attributes={})
      attributes.each_pair { |name, value| instance_variable_set :"@#{name}", value }
      self.children=attributes
    end

    def evaluate(value)
      return nil if not value
      return value.eval if value.is_a? Node
      raise Exception("CANT eval "+x+" "+x.class)
    end

    def interpret
      return children
    end

    def walk(er)
      # yield
      er(self)
      for i in self.children
        next if not i
        i.walk if (i.is_a? Node)
      end
    end

    def dump level=0
      return if not self.children
      print "\t"*level+ self.class.to_s.sub("Kast::", '').sub(/Def$/, '')
      # print "("
      print(":\n")
      z=0
      for k, i in self.children
        next if not i or i.blank?
        if k.to_s=='body'
          level=level-1
        else
          print("\t"*level+" ")
          print(k)
          print("=")

          if i.is_a? String or i.is_a? Numeric
            print i
            print(",") if z>0
            z=z+1
          end
          print("\n")
        end
        next if not i
        i.dump level+1 if (i.is_a? Node)
        i.each do |x|
          x.dump(level+1)
        end if (i.is_a? Array)
        i.each do |x, y|
          if y.is_a?Node
            y.dump(level+1)
          else
            puts "\t"*(level+2)+"#{x}=#{y}"
          end
        end if (i.is_a? Hash)
      end
      return "" #ALREADY DUMPED!!
      # print "\t"*level+")"
      # print("\n")
    end

    def to_s
      type.name
      # dump
    end
  end


# DANGER: CONFLICT if Module not in module (Kast) lol!!
  class Module < Node
    attr_accessor :body

    def interpret
      # return body.map &:interpret
      return body.each { |statement| statement.interpret }
    end
  end

  def Module(body)
    m=Kast::Module.new(body: body)
    # assert(m.body==body)
    m
  end

  class Interactive < Node
    attr_accessor :body
  end

  def Interactive(body)
    Interactive.new({body: body})
  end

  class Expression < Node
    attr_accessor :body
  end

  def Expression(body)
    Expression.new({body: body})
  end


  class Suite < Node
    attr_accessor :body
  end

  def Suite(body)
    Suite.new({body: body})
  end

  class Python3Node < Node
  end

# see class Function !!
  class FunctionDef < Node
    attr_accessor :name, :args, :body, :decorator_list, :returns

    def eval
      define_method(name) do
        super.evaluate(value)
      end
    end
  end

# def FunctionDef(name, args, body=Python3Node, decorator_list, returns)
#   FunctionDef.new({name: name, args: args, body: body, decorator_list: decorator_list, returns: returns})
# end

  def FunctionDef(params={})
    FunctionDef.new(params)
  end

  alias MethodDef FunctionDef
  alias DefDef FunctionDef

  class ClassDef < Node
    attr_accessor :name, :bases, :keywords, :starargs, :kwargs, :body, :decorator_list

    def eval
      # define_class
    end
  end

# def ClassDef(name, bases, keywords=[], starargs=[], kwargs=[], body, decorator_list)
#   ClassDef.new({name: name, bases: bases, keywords: keywords, starargs: starargs, kwargs: kwargs, body: body, decorator_list: decorator_list})
# end
  def ClassDef(params={})
    ClassDef.new(params)
  end

  class Return < Node
    attr_accessor :value

    def eval
      super.evaluate(value)
    end
  end

# def Return(value=nil)
#   Return.new(value: value)
# end

  def Return(params={})
    Return.new(params)
  end


  class Delete < Node
    attr_accessor :targets
  end

  def Delete(targets)
    Delete.new(targets)
  end

  class Assign < Node
    attr_accessor :targets, :value
    alias var targets
    alias vars targets
    alias body value
  end

  # def Assign(targets, value)
  #   Assign.new({targets: targets, value: value})
  # end
  def Assign(params={})
    Assign.new(params)
  end

  # x += 8
  # x *= 8
  class AugAssign < Node
    attr_accessor :targets, :op, :value
  end

  def AugAssign(targets, op, value)
    AugAssign.new({targets: targets, op: op, value: value})
  end

  class While < Node
    attr_accessor :test, :body, :orelse
  end

  def While(test, body, orelse)
    While.new({test: test, body: body, orelse: orelse})
  end

  class If < Node
    attr_accessor :test, :body, :orelse
    alias condition test
    alias then body
    alias else body #DANGER!!
  end

# def If(test, body, orelse)
#   If.new({test: test, body: body, orelse: orelse})
# end
  def If(params={})
    params[:orelse]||=params[:else] # DANGER KEYWORD!
    If.new(params)
  end

  class For < Node
    attr_accessor :target, :iter, :body, :orelse
  end

  def For(target, iter, body, orelse)
    For.new({target: target, iter: iter, body: body, orelse: orelse})
  end


  class With
    attr_accessor :items, :body
  end

  def With(items, body)
    With.new({items: items, body: body})
  end


  class Raise < Node
    attr_accessor :exc, :cause
  end

  def Raise(exc=nil, cause=nil)
    Raise.new({exc: exc, cause: cause})
  end

  class Try < Node
    attr_accessor :body, :handlers, :orelse, :finalbody
  end

  def Try(body, handlers, orelse, finalbody)
    Try.new({body: body, handlers: handlers, orelse: orelse, finalbody: finalbody})
  end

  class Assert < Node
  end

  def Assert(test, msg='')
    Assert.new({test: test, msg: msg})
  end


  class Import < Node
    attr_accessor :names

    def interpret
      return body.map &:interpret
    end
  end

  def Import(names)
    Import.new({names: names}) #.names=names
  end

  class ImportFrom < Node
  end

  def ImportFrom(module_=nil, names=nil, level=0)
    ImportFrom.new({module_: module_, names: names, level: level})
  end

  class None<NilClass;
  end
  class True<TrueClass;
  end
  class False<FalseClass;
  end
# True =true
# None =nil
# False=false
# False=FalseClass

  class Global < Node
    attr_accessor :names
  end

  def Global(names)
    Global.new({names: names}) #.names=names
  end

  class Nonlocal < Node
    attr_accessor :names
  end

  def Nonlocal(names)
    Nonlocal.new({names: names}) #.names=names
  end


  class Expr < Node
    attr_accessor :value
  end

  def Expr(value)
    Expr.new({value: value})
  end

  class Pass<Node
  end

  def Pass
    Pass.new
  end

  class Break<Node
  end

  def Break
    Break.new
  end

  class Continue<Node
  end

  def Continue
    Continue.new
  end


  class BoolOp < Node
    attr_accessor :op, :values
  end

  def BoolOp(op, values)
    BoolOp.new({op: op, values: values})
  end

  class BinOp < Node
    attr_accessor :left, :op, :right

    def interpret
      left.interpret.send(op.to_s, right.interpret)
    end
  end

  def BinOp(left, op, right)
    BinOp.new({left: left, op: op, right: right})
  end

  class UnaryOp < Node
    attr_accessor :op, :operand
  end

  def UnaryOp(op, operand)
    UnaryOp.new({op: op, values: operand})
  end

  class Lambda < Node
    attr_accessor :args, :body
  end

  def Lambda(args, body)
    Lambda.new({args: args, body: body})
  end

# != IF !!!
  class IfExp < Node
    attr_accessor :test, :body, :orelse
  end

  def IfExp(test, body, orelse)
    IfExp.new({test: test, body: body, orelse: orelse})
  end


  class Dict < Node
    attr_accessor :keys, :values
  end

  def Dict(keys, values)
    Dict.new({keys: keys, values: values})
  end


# DANGER: CONFLICT if Set not in module (Kast)
  class Set < Node # NOT ASSIGN!!!!!!!!!!
    attr_accessor :elts

    def initialize(elts)
      self.elts=elts
    end
  end

  def Set(elts)
    # Set.new({elts:elts})
    Set.new(elts)
  end

  alias Let Set

  class SetComp < Node
    attr_accessor :elt, :generators
  end

  def SetComp(elt, generators)
    SetComp.new({elt: elt, generators: generators})
  end

  class ListComp < Node
    attr_accessor :elt, :generators
  end

  def ListComp(elt, generators)
    ListComp.new({elt: elt, generators: generators})
  end

  class DictComp < Node
    attr_accessor :keys, :values, :generators
  end

  def DictComp(keys, values, generators)
    DictComp.new({keys: keys, values: values, generators: generators})
  end

  class GeneratorExp < Node
    attr_accessor :elt, :generators
  end

  def GeneratorExp(elt, generators)
    GeneratorExp.new({elt: elt, generators: generators})
  end


  class Yield < Node
    attr_accessor :value
  end

  def Yield(value=nil)
    Yield.new(value)
  end

  class YieldFrom< Node
    attr_accessor :value
  end

  def YieldFrom(value)
    YieldFrom.new(value)
  end

  class Compare< Node
    attr_accessor :left, :ops, :comparators
  end

  def Compare(left, ops, comparators)
    Compare.new({left: left, ops: ops, comparators: comparators})
  end


  class Call < Node
    attr_accessor :func, :args, :keywords, :starargs, :kwargs
    alias method func
    alias function func
    alias arg args
    alias arguments args
    alias argument args

  end
# def Call(func, args, keywords, starargs=[], kwargs=[], body, decorator_list)
#   Call.new({func: func, args: args, keywords: keywords, starargs: starargs, kwargs: kwargs})
# end

  def Call(params={})
    # params[:keywords]||=[]
    Call.new(params)
  end

  alias Construct Call
# alias Constructor Call


  class Num < Node
    attr_accessor :n

    def interpret
      n.to_i # or to_f ??
    end
  end

  def Num(n)
    Num.new(n: n)
  end

  class Str < Node
    attr_accessor :s

    def interpret
      s
    end
  end

  def Str(s)
    Str.new(s: s)
  end

  class Bytes < Node
    attr_accessor :s
  end

  def Bytes(s)
    Bytes.new(s: s)
  end

  class NameConstant< Node
    attr_accessor :value
  end

  def NameConstant(value)
    NameConstant.new({value: value})
  end

  class Ellipsis < Node
  end

  def Bytes()
    Bytes.new()
  end


  class Attribute< Node
    attr_accessor :value, :attr, :expr_context
  end

  def Attribute(value, attr, expr_context)
    Attribute.new({value: value, attr: attr, expr_context: expr_context})
  end

  class Subscript< Node
    attr_accessor :value, :slice, :ctx
  end

  def Subscript(value, slice, ctx)
    Subscript.new({value: value, slice: slice, ctx: ctx})
  end

  class Starred< Node
    attr_accessor :value, :ctx
  end

  def Starred(value, ctx)
    Starred.new({value: value, ctx: ctx})
  end

  class Name< Node
    attr_accessor :id, :ctx
  end

  def Name(id, ctx)
    Name.new({id: id, ctx: ctx})
  end


  class List< Node
    attr_accessor :elts, :ctx
  end

  def List(elts, ctx)
    List.new({elts: elts, ctx: ctx})
  end

  class Tuple< Node
    attr_accessor :elts, :ctx
  end

  def Tuple(elts, ctx)
    Tuple.new({elts: elts, ctx: ctx})
  end

# class None< NilClass;
# end

  class ExceptionContext <Node;
  end
  class Load < ExceptionContext;
  end
  class Store < ExceptionContext;
  end
  class Del < ExceptionContext;
  end
  class AugLoad < ExceptionContext;
  end
  class AugStore < ExceptionContext;
  end
  class Param < ExceptionContext;
  end

  def Load()
    ; Load.new
  end

  def Store()
    ; Store.new
  end

  def Del()
    ; Del.new
  end

  def AugLoad()
    ; AugLoad.new
  end

  def AugStore()
    ; AugStore.new
  end

  def Param()
    ; Param.new
  end


# alias = (identifier name, identifier? asname)
  class Alias < Node
    attr_accessor :names, :asname
  end

  def Alias(name, asname=nil)
    Alias.new({name: name, asname: asname}) #.names=names
  end

  class Index< Node
    attr_accessor :value
  end
#           Index(expr value)
  def Index(value)
    Index.new(value: value)
  end


  class Slice< Node
    attr_accessor :lower, :upper, :step
  end

  def Slice(lower, upper, step)
    Slice.new(lower: lower, upper: upper, step: step)
  end

  class ExtSlice< Node
    attr_accessor :dims
  end

  def ExtSlice(dims)
    ExtSlice.new({dims: dims})
  end

  class And< BoolOp;
  end
  class Or< BoolOp;
  end

  def And;
    And.new
  end

  def Or;
    Or.new
  end

  class Operator <Node
  end
  class Add < Operator
    def to_s
      '+'
    end
  end
  class Sub < Operator
    def to_s
      '-'
    end
  end
  class Mult < Operator
    def to_s
      '*'
    end
  end
  class Div < Operator
  end
  class Mod < Operator
  end
  class Pow < Operator
  end
  class LShift < Operator
  end
  class RShift < Operator
  end
  class BitOr < Operator
  end
  class BitXor < Operator
  end
  class BitAnd < Operator
  end
  class FloorDiv < Operator
  end

  def Add;
    Add.new
  end

  def Sub;
    Sub.new
  end

  def Mult;
    Mult.new
  end

  def Div;
    Div.new
  end

  def Mod;
    Mod.new
  end

  def Pow;
    Pow.new
  end

  def LShift;
    LShift.new
  end

  def RShift;
    RShift.new
  end

  def BitOr;
    BitOr.new
  end

  def BitXor;
    BitXor.new
  end

  def BitAnd;
    BitAnd.new
  end

  def FloorDiv;
    FloorDiv.new
  end

  class Invert < UnaryOp;
  end
  class Not < UnaryOp;
  end
  class UAdd < UnaryOp;
  end
  class USub < UnaryOp;
  end

  class ComparatorOp < Node;
  end
  class Eq < ComparatorOp;
  end
  class NotEq < ComparatorOp;
  end
  class LtE < ComparatorOp;
  end
  class GtE < ComparatorOp;
  end
  class Is < ComparatorOp;
  end
  class IsNot < ComparatorOp;
  end
  class In < ComparatorOp;
  end
  class NotIn < ComparatorOp;
  end

  def Eq;
    Eq.new
  end

  def NotEq;
    NotEq.new
  end

  def LtE;
    LtE.new
  end

  def GtE;
    GtE.new
  end

  def Is;
    Is.new
  end

  def IsNot;
    IsNot.new
  end

  def In;
    In.new
  end

  def NotIn;
    NotIn.new
  end

# comprehension = (expr target, expr iter, expr* ifs)


  class ExceptHandler < Node
    attr_accessor :type, :name, :body
  end

  def ExceptHandler(type, name, body)
    ExceptHandler.new({type: type, name: name, body: body})
  end

# attributes (int lineno, int col_offset)

  class Arguments < Node
    attr_accessor :args, :vararg, :kwonlyargs, :kw_defaults, :kwarg, :defaults
  end

  def arguments (args, vararg=nil, kwonlyargs=[], kw_defaults=[], kwarg=nil, defaults=nil)
    Arguments.new(args: args, vararg: vararg, kwonlyargs: kwonlyargs, kw_defaults: kw_defaults, kwarg: kwarg, defaults: defaults)
  end

# arguments = (arg* args, arg? vararg, arg* kwonlyargs, expr* kw_defaults,  arg? kwarg, expr* defaults)

# arg = (identifier arg, expr? annotation)
  class Arg< Node
    attr_accessor :arg, :annotation
  end

  def arg(arg, annotation)
    Arg.new(arg: arg, annotation: annotation)
  end


# -- import name with optional 'as' alias.
# alias = (identifier name, identifier? asname)

# -- keyword arguments supplied to call
  class Withitem< Node
    attr_accessor :context_expr, :optional_vars
  end

  def withitem(context_expr, optional_vars)
    Withitem.new({context_expr: context_expr, value: optional_vars})
  end


# -- keyword arguments supplied to call
  class Keyword < Node
    attr_accessor :arg, :value
  end

  def keyword(arg, value)
    Keyword.new({arg: arg, value: value})
  end


# self.value=Call(func=Name(id='print',ctx=Load()), args=values,keywords=[], starargs=None, kwargs=None)
  class Print < Node
    attr_accessor :dest, :values, :nl

    def interpret
      puts(values.interpret)
      # print(values.interpret)
    end
  end

  def Print dest=nil, values=nil, nl=True
    Print.new dest: dest, values: values, nl: nl
  end

  class Condition<Expression
  end

  def Condition(body)
    Condition.new({body: body})
  end

  class Value<Expression
  end

  def Value(body)
    Value.new({body: body})
  end


end
